core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,156 @@
+package read
+
+import (
+	"context"
+	"fmt"
+	"iter"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+type EventQuery struct {
+	Filter           query.KeyFilter
+	EventBinding     *EventBinding
+	SequenceDataType any
+	IsValuePtr       bool
+	Address          common.Address
+}
+
+func MultiEventTypeQuery(ctx context.Context, lp logpoller.LogPoller, eventQueries []EventQuery, limitAndSort query.LimitAndSort) (sequences iter.Seq2[string, commontypes.Sequence], err error) {
+	defer func() {
```

Comment:
Nit: this is a bit too magical for my taste since we're deferring a call AND capturing some values in a closure, and those values may or may not have been modified by the time the defer executes...

Can we simplify this with a helper that we call explicitly at each error return?

Reply:
sure, there are a couple of other places in chainreader that do it this, I'll update those as well

Reply:
there are a couple of places in chainreader that do this, I can sort of see why as its ensures any error always gets the required info attached and not accidentally missed if code is modified and a new error added.  Hmm.  I've updated it with helper, but am wondering if chain reader approach if best.   In this particular context 'those values may or may not have been modified by the time the defer executes' - don't think this is a concern given the state it uses.  Maybe there is a better approach, will give it some thought.